File: benchmarks/jacobi_2d/naive.jl
include("../../timing/dphpc_timing.jl")
using Printf


function init_arrays(n)
    A = zeros(Float64, n, n)
    B = zeros(Float64, n, n)

    A = [((i - 1) * (j + 1) + 2) / n for i in 1:n, j in 1:n]
    B = [((i - 1) * (j + 2) + 3) / n for i in 1:n, j in 1:n]

    return A, B
end


function kernel_j2d(tsteps, n, A, B)
    for t in 1:tsteps
        for i in 2:(n-1)
            for j in 2:(n-1)
                B[i, j] = 0.2 * (A[i, j] + A[i, j-1] + A[i, j+1] + A[i+1, j] + A[i-1, j])
            end
        end
        for i in 2:(n-1)
            for j in 2:(n-1)
                A[i, j] = 0.2 * (B[i, j] + B[i, j-1] + B[i, j+1] + B[i+1, j] + B[i-1, j])
            end
        end
    end
end


function print_array(A)
    #display(stdout, round.(A; digits=2))
    #show(stdout, "text/plain", round.(A; digits=2))
    for i in 1:size(A, 1)
        for j in 1:size(A, 2)
            @printf("%.2f ", A[i, j])
        end
        println()
    end
end

include("_main_cpu.jl")

main()

File: benchmarks/jacobi_2d/naive.c
#include "../../timing/dphpc_timing.h"

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>

#define ASSERT 0

// grid has to be square
void init_arrays(int n, double A[n][n], double B[n][n])
{
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            A[i][j] = ((double) i * (j + 2) + 2) / n; // TODO this is not randomly generated, so we might be optimising for this very specific case instead of generally -> consider
            B[i][j] = ((double) i * (j + 3) + 3) / n;
        }
    }
}

void print_arrays(int n, double A[n][n], double B[n][n])
{   
    puts("matrix A:");
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            printf("%.2f ", A[i][j]);
        }
        printf("\n");
    }
    
    puts("\nmatrix B:");
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            printf("%.2f ", B[i][j]);
        }
        printf("\n");
    }
}

void kernel_j2d(int tsteps, int n, double A[n][n], double B[n][n])
{
    // iterate for tsteps steps
    for (int t = 0; t < tsteps; t++)
    {
        for (int i = 1; i < (n - 1); i++)
        {
            for (int j = 1; j < (n - 1); j++)
            {
                B[i][j] = 0.2 * (A[i][j] + A[i][j-1] + A[i][j+1] + A[i+1][j] + A[i-1][j]); // mean (sum/5) of -|- w/ Aij in centre
            }
        }
        for (int i = 1; i < (n - 1); i++)
        {
            for (int j = 1; j < (n - 1); j++)
            {
                A[i][j] = 0.2 * (B[i][j] + B[i][j-1] + B[i][j+1] + B[i+1][j] + B[i-1][j]);
            }
        }
    }
}

void run_bm(int tsteps, int n, const char* preset)
{    
    double (*A)[n][n] = malloc(sizeof(*A));
    double (*B)[n][n] = malloc(sizeof(*B));

    init_arrays(n, *A, *B);
    
    dphpc_time3(
        init_arrays(n, *A, *B),
        kernel_j2d(tsteps, n, *A, *B),
        preset
    );
    
    if (ASSERT && strcmp(preset, "S") == 0)
    {
        print_arrays(n, *A, *B);
    }

    free((void *) A);
    free((void *) B);
}

int main(int argc, char** argv)
{
    run_bm(50, 150, "S");   // steps 50, n 150
    run_bm(80, 350, "M");   // steps 80, n 350
    run_bm(200, 700, "L");   // steps 200, n 700
    //run_bm(500, 1400, "paper"); // in-between for testing
    run_bm(1000, 2800, "paper");  // steps 1000, n 2800
  
    return 0;
}